### PR TITLE
POC: Use Options, Domain Driven Design entities and Validation library for null-checks

### PR DESCRIPTION
**Pull request by lsolano**
# Maintainers
This is part of a code samples prepared for a conference, not intended for production. If you consider some aspects of the proposal fell free to integrate it, copy code patterns or use it in any way you like.

POC replacing traditional IF based control blocks with Optional types, projections (Map) and Pattern Matching (Match and When). Also replacing manual  null checks for parameters with a library.

0. `MoneyConverter.cs`
  - Factor-out money conversion logic from `CurrencyService.cs`

1. `CurrencyServiceTests.cs` : 
  - Add more test cases to be sure new implementation is working correctly
  - Add negative tests (expecting `NopException` for Zero rates)
  
2. Nop.Services.csproj
  - Added Nuget packages: `Optional` and `Triplex.Validations`

3. `CurrencyService.cs`
  - Convert to file scoped-namespace
  - Re-implemented methods using `MoneyConverter` entity
    - `Task<decimal> ConvertCurrencyAsync(decimal amount, Currency sourceCurrencyCode, Currency targetCurrencyCode)`
    - `Task<decimal> ConvertToPrimaryExchangeRateCurrencyAsync(decimal amount, Currency sourceCurrencyCode)`
    - `Task<decimal> ConvertFromPrimaryExchangeRateCurrencyAsync(decimal amount, Currency targetCurrencyCode)`
  - Introduced some helper methods to reduce duplication
    - `Task<Option<Currency>> GetPrimaryCurrencyAsync()`
    - `Task<Option<Currency>> GetCurrencyByIdAsyncOrNoneAsync(int currencyId)`
    - `decimal ThrowExchangeRateNotFound(Currency currency)`
    - `decimal ThrowPrimaryExchangeRateNotFound()`
    
## Environment Issues

1. `CommonModelFactoryTests.cs`
  - Replace `"\r\n"` with `Environment.NewLine` in `CanPrepareRobotsTextFile()` test case, test was broken on Linux.

### CODE REVIEW

File: src/Libraries/Nop.Services/Directory/CurrencyService.cs [hunk 1]
```diff
@@ -6,296 +6,282 @@
...
+    /// The task result contains the currency
+    /// </returns>
+    public virtual async Task<Currency> GetCurrencyByCodeAsync(string currencyCode)
+    {
+        if (string.IsNullOrEmpty(currencyCode))
+            return null;
 
-        #region Ctor
+        return (await GetAllCurrenciesAsync(true))
+            .FirstOrDefault(c => c.CurrencyCode.ToLowerInvariant() == currencyCode.ToLowerInvariant());
```
Comment (UsamaShahid69) [inline src/Libraries/Nop.Services/Directory/CurrencyService.cs:84]: I think, rather than using ToLowerInvariant() you should make it 
`currencyCode.Equals(c.CurrencyCode, StringComparison.InvariantCulture)`

**Comment by RomanovM** [comment]
@lsolano @UsamaShahid69 Why are you making these changes in the main repository and not in your forks?

**Comment by lsolano** [comment]
@RomanovM Commits from @UsamaShahid69 were added by me, in my fork, by mistake, those are not related to my changes. 

In my fork, the branch is Proof Of Concept (POC), please forgive if you are not interested in those changes, but note that **I fixed a Unit Test around the components that loads the Robot.txt** file that was broken for non-windows OSs due to Line Ending issues, see `src/Tests/Nop.Tests/Nop.Web.Tests/Public/Factories/CommonModelFactoryTests.cs`

**Comment by RomanovM** [comment]
@lsolano Got it, thanks. Please create a new pr for your suggested changes.

**Comment by UsamaShahid69** [comment]
@lsolano  @RomanovM  I am sorry for the inconvenience, I am very new to this. sorry for the trouble I caused.
